### PR TITLE
ORACLEDB: Updated handling of connection errors for disposal

### PR DESCRIPTION
**Pull request by cemremengu**
Solves #2514 and #2603. I tested it on my-machine 😄 I am thinking that simulating this will be tricky and probably take about 15-20 minutes for oracle (or maybe I am exaggerating)

In addition to updating connection error list, we also need to update `oracledb` to check for connection errors.

Disclaimer: I am not good with vanillajs (prototypes etc.) so it would be great if someone reviews the syntax/approach.

### CODE REVIEW

**Comment by elhigu** [comment]
To me code looks pretty good, I would need to checkout the connection.__knex__disposed setting if that is the thing that should be done there (probably it is correct though). @atiertant how does tihs looks to you?

File: src/dialects/oracledb/index.js [hunk 1]
```diff
@@ -135,13 +135,19 @@ Client_Oracledb.prototype.acquireRawConnection = function() {
         if (options.resultSet) {
           connection.execute(sql, bindParams || [], options, function(err, result) {
             if (err) {
+              if (Client_Oracle.prototype.isConnectionError(err)) {
```
Comment (atiertant) [inline src/dialects/oracledb/index.js]: doesn't client.isConnectionError reference the same fonction ?

**Comment by atiertant** [comment]
@cemremengu oracle database has a dead connection detection mechanism but we should call connection.close before set the connection.__knex__disposed.
@elhigu looks great, but didn't tested it

**Comment by cemremengu** [comment]
@atiertant calling `connection.close` seems to throw an error though so I am not sure if it achieves anything
```
 ERROR 8672 nodejs.unhandledRejectionError: NJS-003: invalid connection
```

**Comment by atiertant** [comment]
@cemremengu see node-oracledb examples
https://github.com/oracle/node-oracledb/blob/master/examples/select1.js#L73
should ask @cjbj on a node-oracledb issue which errors needs to close the connection and which other need only to remove the connection object from pool ...
can't find this kind of mechanism in the node-oracledb driver pool, does it has the same bug ?

**Comment by cemremengu** [comment]
Added connection.close() to conform what @cjbj stated

> Closing the connection is the way to remove it from the connection pool, which is important so the pool slot can be reused by a new valid connection. You should always call conn.close(). What you may want to do is ignore errors on the close.

File: src/dialects/oracle/index.js [hunk 2]
```diff
@@ -179,16 +179,22 @@ assign(Client_Oracle.prototype, {
       default:
         return response;
     }
-  }
+  },
 
+  // If the error is any of these, we'll assume we need to
+  // mark the connection as failed
+  isConnectionError(err) {
```
Comment (cjbj) [inline src/dialects/oracle/index.js]: See https://github.com/oracle/odpi/blob/master/src/dpiError.c#L65 for a bigger error list used by the node-oracledb connection pool.  Node-oracledb also uses an OCI call at https://github.com/oracle/odpi/blob/29968fc2085941c40fce8f56a42c9920305d2e29/src/dpiConn.c#L204 which does most of the same checks anyway.

File: src/dialects/oracle/index.js [hunk 3]
```diff
@@ -179,16 +179,22 @@ assign(Client_Oracle.prototype, {
...
+    return [
+      'ORA-03114', // not connected to ORACLE
+      'ORA-03113', // end-of-file on communication channel
+      'ORA-03135', // connection lost contact
+      'ORA-12514', // listener does not currently know of service requested in connect descriptor
+      'NJS-040',
+      'NJS-024',
+      'NJS-003',
+      'NJS-024',
```
Comment (cjbj) [inline src/dialects/oracle/index.js]: NJS-024 is duplicated a few lines above.  And arguably shouldn't be on the list.

**Comment by cjbj** [comment]
The issue heading mentions oracledb but the files changed are for dialect/oracle, not dialect/oracledb.  Is this what you intended?

Also, my comments about connection pooling are for the node-oracledb connection pool, which isn't used in dialects/oracledb

**Comment by kibertoad** [comment]
@cjbj Does that mean that currently used list of errors that should close connections is invalid in case of not using oracledb connection pool?

**Comment by cjbj** [comment]
@kibertoad the ORA errors that indicate a bad connection are pretty much going to be the same for pooled & non-pooled connections.  You could check for the same numbers that ODPI-C uses, see my earlier link.

**Comment by kibertoad** [comment]
@cemremengu Could you please provide me write permissions to your repo so that I could fix the PR accordingly to cjbj's comments without forking and resubmitting the pr?

**Comment by kibertoad** [comment]
@cjbj Thanks!

**Comment by cemremengu** [comment]
@kibertoad Sorry, snowed under work lately. Updated the error list accordingly

**Comment by kibertoad** [comment]
@cemremengu looks like there are random failures, can you rerun tests by closing and opening pr?

**Comment by cemremengu** [comment]
@kibertoad done. Also given you permission in case you want to tweak stuff

**Comment by kibertoad** [comment]
awesome, thank you! also tests passed, yay :)

**Comment by elhigu** [comment]
I think this could be tested by mocking/monkeypatching client.isConnectionError method and returning the original implementation after tests.

**Comment by kibertoad** [comment]
@elhigu What should such test assert?

**Comment by elhigu** [comment]
@kibertoad that connection state is set to disposed in all of the changed code paths (test should fail before this change).

**Comment by alanchristensen** [comment]
What's the status of this PR?

**Comment by kibertoad** [comment]
@alanchristensen I could resolve the merge conflict, but would really appreciate it if someone else wrote the test :)

**Comment by pablopen** [comment]
Hi, any updates on this?
Looks like it could solve the issues when oracle connections are consuming one pool's connection and failing each request that fall in that connection.

**Comment by kibertoad** [comment]
@pablopen Any chance you would volunteer to write at least some test for it?

**Comment by pablopen** [comment]
@kibertoad I could try. I'll check elhigu suggestion of mocking the errors.

**Comment by kibertoad** [comment]
@pablopen You are my hero :)

**Comment by elhigu** [comment]
I would still have preferred having some tests for this, now this is waiting for regression to happen if anyone ever touches that code.

**Comment by kibertoad** [comment]
@elhigu In ideal world, yes. However, year passed, and noone did, and I lost all hope to ever get to it myself; and this change is both straightforward enough and useful (as evidenced by already reported problems solved by it).

**Comment by elhigu** [comment]
In this particular case I considered that benefits were bigger than risks of it causing problems later on.

**Pull request merged**